keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,593 @@
+some manual cleanup on downgrade.
+
+(Note that the middle case covers both "upgrading from
+`DualStackAPIServer=false` to `DualStackAPIServer=true`" _and_
+"upgrading the cluster from single stack to dual stack (after
+`DualStackAPIServer` goes GA)". In both cases, some apiservers will
+begin advertising dual-stack endpoints before other apiservers are
+aware of the configuration change, but the un-upgraded apiservers will
+allow the upgrade to proceed smoothly.)
+
+[kubernetes #101070]: https://github.com/kubernetes/kubernetes/issues/101070
+
+### Test Plan
+
+- We will need unit or integration tests of the behavior under skewed
+  configurations, as described in ["Service and EndpointSlice
+  Reconciling"](#service-and-endpointslice-reconciling). (These tests
+  apply both to the present-day case of "enabling the feature gate in
+  a cluster which is already dual stack" and the future case of
+  "changing the configuration to dual stack in a cluster which already
+  has the feature gate enabled", so they'll still be needed even
+  post-GA.)
+
+- E2E tests of dual-stack apiserver functionality:
+
+    - If the feature is enabled and the cluster is single-stack, then
+      the Service and EndpointSlices should be single-stack
+
+    - If the feature is enabled and the cluster is dual-stack, then
+      the Service and EndpointSlices should be dual-stack, and both
+      the IPv4 and IPv6 service IPs should accept connections.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Feedback from users
+
+- Unit/integration and e2e tests as described above.
+
+- Confirm that CoreDNS behaves as expected (ie, it is not mistakenly
+  special-casing `kubernetes.default` as being obligatorily
+  single-stack).
+
+- We must be alpha for at least 2 releases, to ensure we don't need to
+  support upgrades from pre-Alpha to Beta.
+
+#### Beta -> GA Graduation
+
+- Feedback from users
+
+- 2 examples of distributors/developers/providers making use of
+  dual-stack Endpoints
+
+### Upgrade / Downgrade Strategy
+
+As discussed under ["Service and EndpointSlice
+Reconciling"](#service-and-endpointslice-reconciling), we will modify
+the apiserver behavior when the feature gate is off to make it
+cooperate better with apiservers where the feature gate is on (while
+still having no change in behavior in a cluster where _all_ apiservers
+have the feature gate off). Assuming that the feature remains alpha /
+disabled-by-default for at least 2 releases, this should ensure that
+all non-early-adopters get smooth upgrades and downgrades.
+
+For early adopters, or if we want to only have the feature be alpha
+for a single release, and not have completely smooth behavior for
+users with N-2 version skew, there are a few potential problems:
+
+#### Handling of the "`kubernetes`" Service
+
+kube-apiserver attempts to force the "`kubernetes`" Service definition
+to match its expectations on startup, but does not attempt to
```

Comment:
I think this "does not reconcile" is a bug and when we eventually fix that, it will flap.  I was thinking about alternatives...

EPSlice already says that a given IP *might* be in more than one slice (e.g. while rebalancing).  What if EACH apiserver wrote their OWN epslices.  Single-stack servers would write 1, dual-stack would write 2.   Cleanup on downgrade is tricky - which slices are "mine" ?  Maybe we need some other way to ident apiserver replicas so they can operate independently in this manner?

Reply:
should we create a name convention for the slices name? 
that will simplify the problem `kubernetes` and `kubernetes-ipv6` ?

Reply:
> Maybe we need some other way to ident apiserver replicas so they can operate independently in this manner?

yeah, we already have this; the "lease"-based endpoint reconciling. Each apiserver writes its IPs to a special object in etcd, and then each apiserver can regenerate the Endpoints/EndpointSlices based on that